cmd/catalyst-node/catalyst-node.go

```diff
@@ -751,3 +760,228 @@ func queryMistForClosestNodeSource(playbackID, lat, lon, prefix string, source b
+			return []byte("false")
+		}
+
+		if err := claims.Valid(); err != nil {
+			glog.Errorf("Invalid token %v", err)
+			return []byte("false")
+		}
+
+		if playbackID != claims.Subject {
+			glog.Errorf("PlaybackId mismatch")
+			return []byte("false")
+		}
+
+		pubKey = claims.PublicKey
+
+		if time.Now().After(claims.ExpiresAt.Time) {
+			glog.Errorf("Token expired %v", claims.ExpiresAt)
+			return []byte("false")
+		}
+	}
+
+	playbackAccessControl, err := getPlaybackAccessControlInfo(ac, playbackID, pubKey)
+	if err != nil {
+		glog.Errorf("Unable to get playback access control info %v", err)
+		return []byte("false")
+	}
+
+	if playbackAccessControl[pubKey].allow {
+		return []byte("true")
+	}
+
+	return []byte("false")
+}
+
+func getPlaybackAccessControlInfo(ac *PlaybackAccessControl, playbackID, pubKey string) (map[string]*PlaybackAccessControlEntry, error) {
+	if ac.cache[playbackID] == nil ||
+		ac.cache[playbackID][pubKey] == nil ||
+		time.Now().After(ac.cache[playbackID][pubKey].Stale) {
+
+		err := cachePlaybackAccessControlInfo(ac, playbackID, pubKey)
+		if err != nil {
+			return nil, err
+		}
+
+	} else if time.Now().After(ac.cache[playbackID][pubKey].MaxAge) {
+		go cachePlaybackAccessControlInfo(ac, playbackID, pubKey)
+	}
+
+	return ac.cache[playbackID], nil
+}
+
+func cachePlaybackAccessControlInfo(ac *PlaybackAccessControl, playbackID, pubKey string) error {
+	body, err := json.Marshal(PlaybackAccessControlRequest{"jwt", pubKey, playbackID})
+	if err != nil {
+		return err
+	}
+
+	statusCode, maxAge, stale, err := queryGate(ac, body)
+	if err != nil {
+		return err
+	}
+
+	var allow = statusCode == 204
+	var maxAgeTime = time.Now().Add(time.Duration(maxAge) * time.Second)
+	var staleTime = time.Now().Add(time.Duration(stale) * time.Second)
+
+	if ac.cache[playbackID] == nil {
+		ac.cache[playbackID] = make(map[string]*PlaybackAccessControlEntry)
+	}
+
+	ac.cache[playbackID][pubKey] = &PlaybackAccessControlEntry{staleTime, maxAgeTime, allow}
+
+	return nil
+}
+
+var queryGate = func(ac *PlaybackAccessControl, body []byte) (int, int64, int64, error) {
+	req, err := http.NewRequest("POST", ac.gateURL, bytes.NewReader(body))
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := ac.Client.Do(req)
+	if err != nil {
+		return 0, 0, 0, err
+	}
+
+	defer res.Body.Close()
+
+	cacheControlDirectives := strings.Split(res.Header.Get("Cache-Control"), ",")
+
+	var maxAge, stale int64 = 0, 0
+	if len(cacheControlDirectives) == 2 {
+		maxAge, err = strconv.ParseInt(strings.Split(cacheControlDirectives[0], "=")[1], 10, 64)
+		if err != nil {
+			return 0, 0, 0, err
+		}
+
+		stale, err = strconv.ParseInt(strings.Split(cacheControlDirectives[1], "=")[1], 10, 64)
+		if err != nil {
+			return 0, 0, 0, err
+		}
+	}
+
+	return res.StatusCode, maxAge, stale, nil
+}
+
+type PlaybackGateClaims struct {
+	PublicKey string `json:"pub"`
+	*jwt.RegisteredClaims
```

Comment:
This might be safer as a non-pointer just so we don't have the risk of a NRE below.
```suggestion
	jwt.RegisteredClaims
```